typing/rec_check.ml

```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+            Kept _ -> empty
+          | Overridden (_, e) -> expression e
         in
-        Use.join
-          (use (array field env es))
-          (option expression env eo)
+        join [
+          array field es << field_mode;
+          option expression eo << Dereference
+        ]
     | Texp_ifthenelse (cond, ifso, ifnot) ->
-        Use.(join (inspect (expression env cond))
-                (join
-                    (expression env ifso)
-                    (option expression env ifnot)))
+      (*
+        Gc |- c: m[Dereference]
+        G1 |- e1: m
+        G2 |- e2: m
+        ---
+        Gc + G1 + G2 |- if c then e1 else e2: m
+
+      Note: `if c then e1 else e2` is treated in the same way as
+      `match c with true -> e1 | false -> e2`
+      *)
+      join [
+        expression cond << Dereference;
+        expression ifso;
+        option expression ifnot;
+      ]
     | Texp_setfield (e1, _, _, e2) ->
-        Use.(join (inspect (expression env e1))
-              (inspect (expression env e2)))
+      (*
+        G1 |- e1: m[Dereference]
+        G2 |- e2: m[Dereference] (*TODO: why is this a dereference?*)
+        ---
+        G1 + G2 |- e1.x <- e2: m
+      *)
+      join [
+        expression e1 << Dereference;
+        expression e2 << Dereference;
+      ]
     | Texp_sequence (e1, e2) ->
-      Use.(join (discard (expression env e1))
-              (expression env e2))
-    | Texp_while (e1, e2) ->
-      Use.(join (inspect (expression env e1))
-              (discard (expression env e2)))
+      (*
+        G1 |- e1: m[Guard]
+        G2 |- e2: m
+        --------------------
+        G1 + G2 |- e1; e2: m
+
+        Note: `e1; e2` is treated in the same way as `let _ = e1 in e2`
+      *)
+      join [
+        expression e1 << Guard;
+        expression e2;
+      ]
+    | Texp_while (cond, body) ->
+      (*
+        G1 |- cond: m[Dereference]
+        G2 |- body: m[Guard]
+        ---------------------------------
+        G1 + G2 |- while cond do body done: m
+      *)
+      join [
+        expression cond << Dereference;
+        expression body << Guard;
+      ]
     | Texp_send (e1, _, eo) ->
-      Use.(join (inspect (expression env e1))
-              (inspect (option expression env eo)))
+      (*
+        G |- e: m[Dereference]
+        ---------------------- (plus weird 'eo' option)
+        G |- e#x: m
+      *)
+      join [
+        expression e1 << Dereference;
+        option expression eo << Dereference;
+      ]
     | Texp_field (e, _, _) ->
-      Use.(inspect (expression env e))
-    | Texp_setinstvar (_,_,_,e) ->
-        Use.(inspect (expression env e))
-    | Texp_letexception (_, e) ->
-        expression env e
+      (*
+        G |- e: m[Dereference]
+        -----------------------
+        G |- e.x: m
+      *)
+      expression e << Dereference
+    | Texp_setinstvar (pth,_,_,e) ->
+      (*
+        G |- e: m[Dereference]  (*TODO: why is this a dereference?*)
```

Comment:
As with `Texp_setfield`, this really should be a `Dereference`, so we can drop the `TODO`.

Here's a program for the test suite that fails with a segmentation fault if `Dereference` is weakened to `Return`:

```ocaml
    class c = object
      val mutable f = 0.0
      method m =
        let rec x = (f <- y; ()) and y = 2.0 in f
    end
    let _ = print_float (new c)#m
```